sql/sql_parse_com_rpc.cc

```diff
@@ -0,0 +1,242 @@
+#include "sql_parse_com_rpc.h"
+#include "srv_session.h"
+
+#ifndef EMBEDDED_LIBRARY
+
+static const char* RpcRoleAttr = "rpc_role";
+static const char* RpcDbAttr = "rpc_db";
+
+/*
+ * @retval
+ *  false  success to find attribute
+ * @retval
+ *  true  failed to find attribute
+ * */
+static bool check_for_attribute(THD *thd, const char *attr,
+                                std::string& value) {
+  auto entity = thd->query_attrs_map.find(attr);
+  if (entity != thd->query_attrs_map.end())
+  {
+    value = entity->second;
+    return false;
+  }
+
+  return true;
+}
+
+void update_default_session_object(std::shared_ptr<Srv_session> srv_session,
+                                  THD* conn_thd, bool used_default_srv_session)
+{
+  if (srv_session->session_state_changed())
+  {
+    // if the default object was reused for this session, reset it
+    if (used_default_srv_session)
+      conn_thd->set_default_srv_session(nullptr);
+  }
+  else
+  {
+    // store as default so we don't allocate for next not in session query
+    if (!conn_thd->get_default_srv_session())
+    {
+      DBUG_PRINT("info", ("No session change, reuse object"));
+      conn_thd->set_default_srv_session(srv_session);
+    }
+  }
+}
+
+void reset_conn_thd_after_query_execution(THD* thd) {
+  thd->update_server_status();
+
+  thd->reset_query();
+  thd->reset_query_attrs();
+  thd->set_command(COM_SLEEP);
+  thd->proc_info= 0;
+  thd->m_digest= NULL;
+
+  if (thd->m_statement_psi)
+  {
+    MYSQL_END_STATEMENT(thd->m_statement_psi, thd->get_stmt_da());
+    thd->m_statement_psi= NULL;
+  }
+
+  thd->get_stmt_da()->reset_message();
+  thd->packet.shrink(thd->variables.net_buffer_length);	// Reclaim some memory
+  free_root(thd->mem_root,MYF(MY_KEEP_PREALLOC));
+}
+
+/*
+  @retval
+    0  success
+  @retval
+    1  failure
+*/
+bool handle_com_rpc(THD *conn_thd, char* packet, uint packet_length,
+                    size_t attrs_len, bool* is_rpc_query)
+{
+  DBUG_ENTER(__func__);
+  std::string rpc_role, rpc_db, rpc_id;
+  bool used_default_srv_session = false;
+  bool ret = true;
+  Diagnostics_area * stmt_da_to_restore = NULL;
+  Security_context* conn_security_ctx = NULL;
+  Vio* vio_restore = NULL;
+  THD* srv_session_thd = NULL;
+
+  check_for_attribute(conn_thd, RpcRoleAttr, rpc_role);
+  check_for_attribute(conn_thd, RpcDbAttr, rpc_db);
+  check_for_attribute(conn_thd, Srv_session::RpcIdAttr, rpc_id);
+
+  if (!rpc_role.size() && !rpc_id.size())
+  {
+    // not an rpc command;
+    DBUG_PRINT("info", ("Not an RPC query"));
+    *is_rpc_query = false;
+    DBUG_RETURN(true);
+  }
+
+  *is_rpc_query = true;
+
+  DBUG_PRINT("info", ("rpc_role='%s', rpc_db='%s', rpc_id='%s'",
+                      rpc_role.c_str(), rpc_db.c_str(), rpc_id.c_str()));
+
+  std::shared_ptr<Srv_session> srv_session;
```

Comment:
```suggestion
Srv_session_ptr srv_session;
```

Reply:
I am not sure, this hides pointer behind a type alias, leaving just a "_ptr", and it is not obvious from _ptr, what kind of pointer is it. std::shared_ptr<foo> is immediately obvious.